### DISCUSSION TITLE
CUID and UUID with PostgreSQL and Foreign Keys

### DISCUSSION THREAD

**Discussion by LFCavalcanti** [prisma/prisma#21489] [Q&A]
### Question

Hello everyone!

So, I'm having difficulty to setup UUID and CUIDs on my App using Prisma on PostgreSQL.

Currently I have the PK using CUID in some tables to provide better control and scalability, the schema is defined as:
```prisma
model CompanyGroup {
  id              String @id @default(cuid())
}
```

Or in case of more security I'm using UUId's like so:
```prisma
model User {
  id            String    @id @default(uuid())
}
```

I read that UUIDs should be stored in a specific binary type so indexing is done correctly.

Searching Prisma Docs I did found a way to deal with UUIDs with the native type:
https://www.prisma.io/docs/concepts/database-connectors/postgresql#native-type-mappings
https://www.prisma.io/docs/reference/api-reference/prisma-schema-reference#uuid

So I believe the correct model for the User table would be:
```
model User {
  id   String  @id @default(uuid()) @db.Uuid
```

But I didn't find a way to store CUID, I tried using the cuid() function with the @db.Uuid type mapping but it fails because the string structure is different.

Another assumption is that FKs should use the same native type, so in example the session table should be:
```prisma
model Session {
  id           String   @id @default(cuid()) @db.Uuid
  sessionToken String   @unique
  userId       String @db.Uuid
  expires      DateTime
  user         User     @relation(fields: [userId], references: [id], onDelete: Cascade)
}
```

My questions are:

1. How to generate and correctly store CUIDs in PostgreSQL using Prisma, using the correct types and optimize the schema?
2. How to deal with FKs when using these CUID and UUID data types?
3. Am I being to concerned about not just using a varchar column to store these?

### Information about Prisma Schema, Client Queries and Environment (optional)

Prisma 5.4.2
@prisma/client 5.4.2

The app is using Next.js 13.5.4

I just updated all dependencies yesterday 😄

**Comment by ludralph** [comment]
Hi @LFCavalcanti 👋 

> How to generate and correctly store CUIDs in PostgreSQL using Prisma, using the correct types and optimize the schema?

To generate `CUIDs` in Prisma, you can use the `@default(cuid())` attribute in your Prisma schema. However, it's important to note that Prisma 1 `CUIDs` were represented as strings with a maximum size of 25 characters (VARCHAR(25)), but in Prisma 2.x or later versions, the generated `CUIDs` might exceed this limit and can be up to 30 characters long. Therefore, you should adjust the column type to VARCHAR(30) to accommodate the potential length of `CUIDs` generated by Prisma 2.x or later versions. See the [docs](https://www.prisma.io/docs/orm/more/upgrade-guides/upgrade-from-prisma-1/schema-incompatibilities-postgresql#problem-8)
```prisma
model CompanyGroup {
  id String @id @default(cuid()) @db.VarChar(30)
}
```

> How to deal with FKs when using these CUID and UUID data types?

When using foreign keys (FKs) with `CUIDs` or `UUIDs`, you should ensure that the FK columns use the same data type as the primary key (PK) columns they reference. If you're using `@db.Uuid` for `UUIDs`, you should use the same for the FKs that reference `UUIDs`. For `CUIDs` stored as VARCHAR(30), the FKs should also be of type VARCHAR(30).
```prisma
model Session {
  id           String   @id @default(cuid()) @db.VarChar(30)
  sessionToken String   @unique
  userId       String   @db.VarChar(30)
  expires      DateTime
  user         User     @relation(fields: [userId], references: [id], onDelete: Cascade)
}
```
> Am I being too concerned about not just using a varchar column to store these?

Using the correct data type is important for both performance and data integrity. For `UUIDs`, using the native `UUID` type in PostgreSQL can provide better performance for indexing and querying. For `CUIDs`, ensuring that the column size is sufficient to store the generated values is crucial. While using VARCHAR may work, it's important to use the correct length and consider the database's native types when available for optimization.

If this answers your question, it would be great if you could **mark this Discussion as answered** to indicate that it has been resolved. 

Otherwise please let us know how else we can help you further or close the Discussion if it was resolved in some other way 🙏

**Comment by LFCavalcanti** [comment]
Hi @ludralph !

Thanks for the information.

I'm still puzzled with the native binary types and what Prisma currently support or not.

I've read this blog about ULIDs: https://blog.daveallie.com/ulid-primary-keys

There's implementations for UUIDv7 or ULID as custom extensions, but if Prisma doesn't support the types I't a moot point.

**Comment by kibertoad** [comment]
@LFCavalcanti CUID is not format-compatible with UUID, so you can't store it in an optimized UUID format in PostgreSQL, unfortunately. If you use 128-bit ULID format, it is compatible with UUID and could be stored in an optimized UUID format. E. g. in CockroachDB this works:

```
id        String       @id @default(dbgenerated("gen_random_ulid()")) @db.Uuid
```

Now PostgreSQL (as opposed to CockroachDB) doesn't natively support ULID generation, so you would need to install an extension, similar to https://github.com/pksunkara/pgx_ulid, in order to be able to generate them. But if you have the extension, `dbgenerated` should work just fine, same as with the CockroachDB.

**Comment by jayarjo** [comment]
Does Prisma export cuid generation utility? I need it for raw SQL queries.

**Reply by ludralph** [reply]
No we don't export it. For raw SQL queries, you can use the cuid library in your application code to generate IDs before inserting them into the database.
